dfusion_rust_core/src/models/order.rs

```diff
@@ -47,6 +47,73 @@ impl Order {
             sell_amount,
         }
     }
+
+    /// Calculates the utility of an order based on a buy and sell price as well
+    /// as the executed buy amount.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Some(U256)` utility result if the inputs were valid and it
+    /// produced a valid positive utility.
+    pub fn utility(
+        &self,
+        buy_price: u128,
+        exec_buy_amount: u128,
+        exec_sell_amount: u128,
+    ) -> Option<U256> {
+        let buy_price = u128_to_u256(buy_price);
+        let exec_buy_amount = u128_to_u256(exec_buy_amount);
+        let exec_sell_amount = u128_to_u256(exec_sell_amount);
+        let buy_amount = u128_to_u256(self.buy_amount);
+        let sell_amount = u128_to_u256(self.sell_amount);
+
+        // utility = (exec_buy_amount - (exec_sell_amount * buy_amount) / sell_amount)
+        //   * buy_price
+
+        let utility = exec_buy_amount
```

Comment:
Do we need checked operations here? Are there examples where this can overflow?

Reply:
Yes we do need to check. Since we don't actually check if the solution is valid, things like having a higher `exec_sell_amount` than `sell_amount` can lead to overflows.

Reply:
Well, we have a few options:
- assume that the solver always behaves, then we don't have to check anything and can just use regular math. This can leads to sub-optimal gas usage in that we submit solutions that won't actually succeed
- do some checking as with this math

Reply:
Yes, if we check the solution, I'd suggest to do a separate method (something like https://github.com/gnosis/dex-services/blob/87a6826420f3a1410c0749ac2db56199da24a73f/driver/src/price_finding/naive_solver.rs#L604-L608).

For the purpose of utility calculation I'd suggest we assume a valid solution (since we don't check balances etc. anyways).

Reply:
I removed some of the overflow checking where it isn't needed but kept it in some places where it could happen.